src/traces/bar/attributes.js

```diff
@@ -86,6 +86,28 @@ module.exports = {
         ].join(' ')
     },
 
+    insidetextanchor: {
+        valType: 'enumerated',
+        values: ['end', 'middle', 'start'],
+        dflt: 'end',
+        role: 'info',
+        editType: 'plot',
+        description: [
+            'Determines if texts are kept at center or start/end points in `textposition` *inside* mode.'
+        ].join(' ')
+    },
+
+    insidetextrotate: {
+        valType: 'enumerated',
+        values: ['auto', 'none'],
+        dflt: 'auto',
+        role: 'info',
+        editType: 'plot',
+        description: [
+            'Determines if texts could automatically be rotated to fit inside bars in `textposition` *inside* mode.'
+        ].join(' ')
+    },
```

Comment:
We'll need approval from @plotly/plotly_js and @nicolaskruchten for these new bar/waterfall/funnel attributes (perhaps during this week's meeting).

I'm a fan of `insidetextanchor` :ok_hand:  

For `insidetextrotate`, perhaps `textangle` (as proposed in https://github.com/plotly/plotly.js/issues/3117) would be better. If implementing a general `textangle` behavior would take too long, we could start with implementing an _enumerated_ `textangle` with values `auto`, `0` and `90`.

Reply:
For funnels, I'd rather we not do the funky rotation stuff by default :)

Reply:
Oh yes. As discussed for funnels the auto rotation and text anchors of inside texts are set differently from bars and waterfalls: 
https://github.com/plotly/plotly.js/blob/4c741c7f477d53033ec320cbbb52c9c8c1d33f1d/src/traces/funnel/attributes.js#L39-L40

Reply:
Well, IMO all new traces shouldn't do it... is it possible to disable it by default for waterfall also or is it too late? sunburst? :)

Reply:
> shouldn't do it

Shouldn't do what? Auto-Rotate the labels?

Reply:
TBD Thursday :busts_in_silhouette:

Reply:
> Shouldn't do what? Auto-Rotate the labels?

yes. discuss thursday :)

Reply:
General `textangle` for `bar`, `waterfall` and `funnel` addressed in https://github.com/plotly/plotly.js/pull/3817/commits/380db2ecaa750ae317c06e2aaf1d31aac3139bdf. `Funnel` has a different default i.e. `0` not `auto`.

Reply:
Looking good! Thanks for going the extra mile and making `textrotate` work for general angles! Your two new mocks are absolutely brilliant.

Now, I see that you updated the `bar_attrs_relative.png` baseline. Why is that?

Reply:
One more thing, `textangle` for `textposition: 'outside'` looks off. For example:

```js
Plotly.newPlot(gd, [{
  type: 'bar',
  y: [1, 2, 1],
  text: ['Loong', 'Loonggggerr', 'Suuuupeeeer Loooooong'],
  textposition: 'outside',
  textangle: -20,
  cliponaxis: false
}])
```

gives:

![image](https://user-images.githubusercontent.com/6675409/57232334-5c1dd480-6fea-11e9-9dd0-fb608b34542d.png)

I think we should move the text-anchor so the text nodes are fully outside the bars.

----------

Now, I don't want you spending too much time on fixing `textangle` for `textposition: 'outside'` as it isn't part of our requirements for 1.48.0. If you think fixing that will take you more than a few hours, please make `textangle` work only for `textposition: 'inside'` for now and move on. Thanks!

Reply:
Good call. Addressed in https://github.com/plotly/plotly.js/pull/3817/commits/30b172a613626b491b8b814b4a392d45f8b4c681.

Reply:
Still extra padding required for this [codepen](https://codepen.io/MojtabaSamimi/pen/OYPOLd?editors=0010).

Reply:
Now fixed in https://github.com/plotly/plotly.js/pull/3817/commits/446baccaed5712736a21fea85f33078347db0ef5.
[Codepen](https://codepen.io/MojtabaSamimi/pen/joEYxZ?editors=0010)

Reply:
> Now fixed in [446bacc](https://github.com/plotly/plotly.js/commit/446baccaed5712736a21fea85f33078347db0ef5).

Very nice!

Now, what about

> Now, I see that you updated the bar_attrs_relative.png baseline. Why is that?

Reply:
Now resolved!